Lib/socketserver.py

```diff
@@ -651,24 +652,46 @@ def process_request_thread(self, request, client_address):
         except Exception:
             self.handle_error(request, client_address)
         finally:
-            self.shutdown_request(request)
+            try:
+                self.shutdown_request(request)
+            finally:
+                self._remove_thread()
+
+    def _remove_thread(self):
+        """Remove a current thread from threads list."""
+        thread = threading.current_thread()
+        if not thread.daemon:
+            if self._threads_lock is None:
+                self._threads_lock = threading.Lock()
+            with self._threads_lock:
+                if self._threads is not None:
+                    try:
+                        self._threads.remove(thread)
+                    except ValueError:
+                        pass
 
     def process_request(self, request, client_address):
         """Start a new thread to process the request."""
         t = threading.Thread(target = self.process_request_thread,
                              args = (request, client_address))
         t.daemon = self.daemon_threads
         if not t.daemon and self.block_on_close:
-            if self._threads is None:
-                self._threads = []
-            self._threads.append(t)
+            if self._threads_lock is None:
```

Comment:
Would it be possible to initialize _threads_lock and _threads *before* creating the first thread?

Reply:
These are now created _before_ creating the first thread.